test/test_foreach.py

```diff
@@ -118,83 +128,368 @@ def test_addcdiv(self, device, dtype):
+                    foreach_bin_op_(tensors, scalar)
+            else:
+                foreach_bin_op_(tensors, scalar)
+                self.assertEqual(res, tensors)
+
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_complex_scalarlist(self, device, dtype):
+        for foreach_bin_op, foreach_bin_op_, torch_bin_op in zip(self.foreach_bin_ops,
+                                                                 self.foreach_bin_ops_,
+                                                                 self.torch_bin_ops):
+            tensors = self._get_test_data(device, dtype, 3)
+            scalars = [3 + 5j, 3 + 5j, 3 + 5j]
+            expected = [torch_bin_op(t, s) for t, s in zip(tensors, scalars)]
+
+            if dtype == torch.bool:
+                if foreach_bin_op == torch._foreach_sub:
+                    with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator, with two bool"):
+                        foreach_bin_op_(tensors, scalar)
+
+                    with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator, with two bool"):
+                        foreach_bin_op(tensors, scalar)
+                return
+
+            with self.assertRaisesRegex(TypeError, "argument 'scalars' must be tuple of floats"):
+                res = foreach_bin_op(tensors, scalars)
+
+            with self.assertRaisesRegex(TypeError, "argument 'scalars' must be tuple of floats"):
+                foreach_bin_op_(tensors, scalars)
 
     @dtypes(*torch.testing.get_all_dtypes())
     def test_bool_scalar(self, device, dtype):
-        tensors = [torch.zeros(10, 10, device=device, dtype=dtype) for _ in range(10)]
-        bool_scalar = True
+        for foreach_bin_op, foreach_bin_op_, torch_bin_op in zip(self.foreach_bin_ops,
+                                                                 self.foreach_bin_ops_,
+                                                                 self.torch_bin_ops):
+            tensors = self._get_test_data(device, dtype, 3)
+            scalar = True
+
+            if dtype == torch.bool:
+                expected = [torch_bin_op(t, scalar) for t in tensors]
+                res = foreach_bin_op(tensors, scalar)
+
+                foreach_bin_op_(tensors, scalar)
+                self.assertEqual(tensors, res)
+                return
 
-        expected = [torch.ones(10, 10, device=device, dtype=dtype) for _ in range(10)]
+            if foreach_bin_op == torch._foreach_sub and device == "cpu":
+                with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator"):
+                    res = foreach_bin_op(tensors, scalar)
 
-        res = torch._foreach_add(tensors, bool_scalar)
-        self.assertEqual(res, expected)
+                with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator"):
+                    foreach_bin_op_(tensors, scalar)
+            elif foreach_bin_op == torch._foreach_sub and "cuda" in device:
```

Comment:
The difference comes from the fact that in this case (bool scalar on CUDA) we go via 'fast' path to our new Kernel. This kernel will have result of the same dtype as an input and will bool scalar will be converted to dtype as well. 
See [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu#L21)

Example #1
Device: CPU
Scalar: Bool 
Dtype: Bfloat16

It will go via "slow" path where subtraction with bool dtype is forbidden. 

Example #2
Device: CUDA
Scalar: Bool 
Dtype: Bfloat16

We will go via fast path due to [these](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ForeachUtils.h#L40) conditions, Scalar will be converted to bfloat16 and result will be bfloat16